# PosMul ì˜ˆì¸¡ ê²Œì„ - Cursor ê·œì¹™ (í•œê¸€)

ë‹¹ì‹ ì€ TypeScript, Next.js 15 App Router, ë„ë©”ì¸ ì£¼ë„ ì„¤ê³„(DDD), í´ë¦° ì•„í‚¤í…ì²˜, Supabase ì „ë¬¸ê°€ì…ë‹ˆë‹¤.

## í”„ë¡œì íŠ¸ ê°œìš”

PosMulì€ ì˜ˆì¸¡ ê²Œì„ê³¼ ì§€ì—­ ê²½ì œ í†µí•©ì„ ì œê³µí•˜ëŠ” AI ì‹œëŒ€ ì§ì ‘ë¯¼ì£¼ì£¼ì˜ í”Œë«í¼ì…ë‹ˆë‹¤. ì´ í”Œë«í¼ì€ ëŒ€ë¦¬ì¸ ì´ë¡ (Agency Theory), CAPM, í–‰ë™ê²½ì œí•™ ì´ë¡ ì„ ì½”ë“œë¡œ êµ¬í˜„í•©ë‹ˆë‹¤.

**ì¤‘ìš”**: ì´ í”„ë¡œì íŠ¸ëŠ” **ëª¨ë“  ë„ë©”ì¸ì— ê±¸ì¹œ ê¸´ë°€í•˜ê²Œ ê²°í•©ëœ ê²½ì œ ì‹œìŠ¤í…œ(PMP/PMC)**ì„ ê°€ì§€ê³  ìˆìŠµë‹ˆë‹¤. ê²½ì œ í†µí•©ì„ ìœ„í•´ **ê³µìœ  ì»¤ë„(Shared Kernel)**ê³¼ **ë„ë©”ì¸ ì´ë²¤íŠ¸(Domain Events)** íŒ¨í„´ì„ ì‚¬ìš©í•˜ì„¸ìš”.

## ê°œë°œ í™˜ê²½

- **ìš´ì˜ì²´ì œ**: Windows
- **ì…¸**: PowerShell (í•­ìƒ PowerShell ëª…ë ¹ì–´ ì‚¬ìš©)
- **ê²½ë¡œ í˜•ì‹**: Windows ë°±ìŠ¬ë˜ì‹œ ê²½ë¡œ (ì˜ˆ: `src\bounded-contexts\`)
- **ëª…ë ¹ì–´**: ëª…ë ¹ì–´ ì²´ì´ë‹ì— `&&` ëŒ€ì‹  ì„¸ë¯¸ì½œë¡ (`;`) ì‚¬ìš©

## í•µì‹¬ ì•„í‚¤í…ì²˜ ì›ì¹™

### ê²½ì œ í†µí•©ì´ í¬í•¨ëœ DDD + í´ë¦° ì•„í‚¤í…ì²˜

- **ê³µìœ  ì»¤ë„**: ê²½ì œ ë„ë©”ì¸(PMP/PMC)ì€ ëª¨ë“  ì»¨í…ìŠ¤íŠ¸ì—ì„œ ê³µìœ 
- **ë„ë©”ì¸ ì´ë²¤íŠ¸**: ë„ë©”ì¸ ê°„ í†µì‹ ì— ì´ë²¤íŠ¸ ì‚¬ìš©
- **ë¶€íŒ¨ ë°©ì§€ ê³„ì¸µ**: í†µí•©ì„ í—ˆìš©í•˜ë©´ì„œ ë„ë©”ì¸ ë¬´ê²°ì„± ë³´í˜¸
- ë„ë©”ì¸ ê³„ì¸µì€ ì™¸ë¶€ ì˜ì¡´ì„±ì´ ì—†ìŒ (ìˆœìˆ˜ ë¹„ì¦ˆë‹ˆìŠ¤ ë¡œì§)
- Aggregate, Entity, Value Object, Domain Serviceë¥¼ ì ì ˆíˆ ì‚¬ìš©
- ë°ì´í„° ì ‘ê·¼ ì¶”ìƒí™”ë¥¼ ìœ„í•œ Repository íŒ¨í„´

### ì˜ì¡´ì„± ê·œì¹™ (ì¤‘ìš”)
```
í”„ë ˆì  í…Œì´ì…˜ â†’ ì• í”Œë¦¬ì¼€ì´ì…˜ â†’ ë„ë©”ì¸
ì¸í”„ë¼ìŠ¤íŠ¸ëŸ­ì²˜ â†’ ë„ë©”ì¸ (ì¸í„°í˜ì´ìŠ¤ êµ¬í˜„)
ê³µìœ  ê²½ì œ ì»¤ë„ â† ëª¨ë“  ë„ë©”ì¸ (ì½ê¸° ì „ìš© ì ‘ê·¼)
```
- ë„ë©”ì¸ì€ ì ˆëŒ€ ì™¸ë¶€ ê³„ì¸µì— ì˜ì¡´í•˜ì§€ ì•ŠìŒ
- ì• í”Œë¦¬ì¼€ì´ì…˜ì€ ë„ë©”ì¸ê³¼ ê³µìœ  ê²½ì œ ì»¤ë„ì—ë§Œ ì˜ì¡´
- ì¸í”„ë¼ìŠ¤íŠ¸ëŸ­ì²˜ëŠ” ë„ë©”ì¸ ì¸í„°í˜ì´ìŠ¤ë¥¼ êµ¬í˜„
- **ëª¨ë“  ë„ë©”ì¸ì€ ê²½ì œ ì»¤ë„ì—ì„œ ì½ê¸°ë§Œ ê°€ëŠ¥, ì“°ê¸°ëŠ” ë„ë©”ì¸ ì´ë²¤íŠ¸ë¥¼ í†µí•´ì„œë§Œ**

## ê²½ì œ ì‹œìŠ¤í…œ í†µí•© íŒ¨í„´

### 1. ê³µìœ  ì»¤ë„ íŒ¨í„´ (ê²½ì œ ë„ë©”ì¸)
```typescript
// src/shared/economy-kernel/
export interface PmpAccount {
  userId: UserId;
  balance: number;
  readonly transactions: PmpTransaction[];
}

export interface PmcAccount {
  userId: UserId;
  balance: number;
  readonly transactions: PmcTransaction[];
}

export class EconomyKernel {
  // ëª¨ë“  ë„ë©”ì¸ì—ì„œ ì½ê¸° ì „ìš© ì ‘ê·¼
  getPmpBalance(userId: UserId): Promise<number>;
  getPmcBalance(userId: UserId): Promise<number>;
  
  // ì“°ê¸° ì‘ì—…ì€ ë„ë©”ì¸ ì´ë²¤íŠ¸ë¥¼ í†µí•´ì„œë§Œ
  private constructor() {} // ì§ì ‘ ì¸ìŠ¤í„´ìŠ¤í™” ë°©ì§€
}
```

### 2. ê²½ì œ ê±°ë˜ë¥¼ ìœ„í•œ ë„ë©”ì¸ ì´ë²¤íŠ¸
```typescript
// ë„ë©”ì¸ ê°„ ê²½ì œ ì´ë²¤íŠ¸
export class PmpEarnedEvent implements DomainEvent {
  constructor(
    public readonly userId: UserId,
    public readonly amount: number,
    public readonly source: 'major-league' | 'brainstorming' | 'debate',
    public readonly sourceId: string,
    public readonly occurredAt: Date = new Date()
  ) {}
}

export class PmcEarnedEvent implements DomainEvent {
  constructor(
    public readonly userId: UserId,
    public readonly amount: number,
    public readonly source: 'local-league' | 'cloud-funding' | 'prediction-success' | 'gift-aid',
    public readonly sourceId: string,
    public readonly occurredAt: Date = new Date()
  ) {}
}

export class PmpSpentEvent implements DomainEvent {
  constructor(
    public readonly userId: UserId,
    public readonly amount: number,
    public readonly purpose: 'prediction-participation',
    public readonly targetId: string,
    public readonly occurredAt: Date = new Date()
  ) {}
}
```

### 3. ê²½ì œ í†µí•©ì„ ìœ„í•œ ë¶€íŒ¨ ë°©ì§€ ê³„ì¸µ
```typescript
// ê° ë„ë©”ì¸ì€ ìì²´ ê²½ì œ ì„œë¹„ìŠ¤ë¥¼ ê°€ì§
export class PredictionEconomicService {
  constructor(
    private readonly economyKernel: EconomyKernel,
    private readonly eventPublisher: IDomainEventPublisher
  ) {}

  async canParticipateInPrediction(
    userId: UserId,
    requiredPmp: number
  ): Promise<boolean> {
    const balance = await this.economyKernel.getPmpBalance(userId);
    return balance >= requiredPmp;
  }

  async participateInPrediction(
    userId: UserId,
    predictionId: PredictionId,
    pmpAmount: number
  ): Promise<Result<void, EconomicError>> {
    // ì§ì ‘ ìˆ˜ì • ëŒ€ì‹  ì´ë²¤íŠ¸ ë°œí–‰
    await this.eventPublisher.publish(
      new PmpSpentEvent(userId, pmpAmount, 'prediction-participation', predictionId)
    );
    return { success: true, data: undefined };
  }
}
```

## ê²½ì œ í†µí•©ì„ ìœ„í•œ íŒŒì¼ êµ¬ì¡° ê·œì¹™

```
src\
  shared\
    economy-kernel\          # ê³µìœ  ê²½ì œ ë„ë©”ì¸
      entities\
        pmp-account.entity.ts
        pmc-account.entity.ts
        money-wave.entity.ts
      value-objects\
        economic-types.ts
      services\
        economy-kernel.service.ts
      events\
        economic-events.ts
    events\                  # ë„ë©”ì¸ ê°„ ì´ë²¤íŠ¸ ì‹œìŠ¤í…œ
      domain-events.ts
      event-publisher.ts
  bounded-contexts\
    [ì»¨í…ìŠ¤íŠ¸ëª…]\
      domain\
        entities\
        value-objects\
        repositories\
        services\
          [ì»¨í…ìŠ¤íŠ¸]-economic.service.ts  # ë„ë©”ì¸ë³„ ê²½ì œ í†µí•©
        events\
      application\
        use-cases\
        services\
        dto\
        event-handlers\      # ê²½ì œ ì´ë²¤íŠ¸ ì²˜ë¦¬
      infrastructure\
        repositories\
        external-services\
        event-handlers\      # ì¸í”„ë¼ ì´ë²¤íŠ¸ í•¸ë“¤ëŸ¬
      presentation\
        components\
        hooks\
      context.md
```

## ë„ë©”ì¸ë³„ ê²½ì œ í†µí•© ê·œì¹™

### ì˜ˆì¸¡ ë„ë©”ì¸
```typescript
// ì˜ˆì¸¡ ë„ë©”ì¸ ê²½ì œ í†µí•©
export class PredictionGame {
  public participateWithPmp(
    userId: UserId,
    prediction: Prediction,
    economicService: PredictionEconomicService
  ): Promise<Result<void, BusinessRuleError>> {
    // 1. ê²½ì œ ì„œë¹„ìŠ¤ë¥¼ í†µí•´ PMP ì”ì•¡ í™•ì¸
    // 2. ì˜ˆì¸¡ ìƒì„±
    // 3. PmpSpentEvent ë°œí–‰
    // 4. ì˜ˆì¸¡ ì„±ê³µ ì‹œ PmcEarnedEvent ë°œí–‰
  }
}
```

### íˆ¬ì ë„ë©”ì¸
```typescript
// íˆ¬ì ë„ë©”ì¸ ê²½ì œ í†µí•©
export class MajorLeague {
  public completeInvestment(
    userId: UserId,
    investment: Investment
  ): Promise<Result<void, BusinessRuleError>> {
    // ë©”ì´ì € ë¦¬ê·¸ ì™„ë£Œ ì‹œ PmpEarnedEvent ë°œí–‰
  }
}

export class LocalLeague {
  public completeInvestment(
    userId: UserId,
    investment: Investment
  ): Promise<Result<void, BusinessRuleError>> {
    // ë¡œì»¬ ë¦¬ê·¸ ì™„ë£Œ ì‹œ PmcEarnedEvent ë°œí–‰
  }
}
```

### í¬ëŸ¼ ë„ë©”ì¸
```typescript
// í¬ëŸ¼ ë„ë©”ì¸ ê²½ì œ í†µí•©
export class Debate {
  public completeDebate(
    userId: UserId,
    debate: Debate
  ): Promise<Result<void, BusinessRuleError>> {
    // í† ë¡  ì°¸ì—¬ ì‹œ PmpEarnedEvent ë°œí–‰
  }
}

export class Brainstorming {
  public completeBrainstorming(
    userId: UserId,
    session: BrainstormingSession
  ): Promise<Result<void, BusinessRuleError>> {
    // ë¸Œë ˆì¸ìŠ¤í† ë° ì°¸ì—¬ ì‹œ PmpEarnedEvent ë°œí–‰
  }
}
```

### ê¸°ë¶€ ë„ë©”ì¸
```typescript
// ê¸°ë¶€ ë„ë©”ì¸ ê²½ì œ í†µí•©
export class Donation {
  public makeDonation(
    userId: UserId,
    amount: number,
    target: DonationTarget,
    economicService: DonationEconomicService
  ): Promise<Result<void, BusinessRuleError>> {
    // 1. PMC ì”ì•¡ í™•ì¸
    // 2. ê¸°ë¶€ ìƒì„±
    // 3. PmcSpentEvent ë°œí–‰
  }
}
```

## ê²½ì œ ì´ë²¤íŠ¸ ì²˜ë¦¬ íŒ¨í„´

### ê²½ì œ ë„ë©”ì¸ì˜ ì´ë²¤íŠ¸ í•¸ë“¤ëŸ¬
```typescript
// ê²½ì œ ë„ë©”ì¸ì´ ëª¨ë“  ê²½ì œ ì´ë²¤íŠ¸ ì²˜ë¦¬
export class PmpAccountEventHandler {
  async handle(event: PmpEarnedEvent | PmpSpentEvent): Promise<void> {
    // PMP ê³„ì • ì”ì•¡ ì—…ë°ì´íŠ¸
    // ê±°ë˜ ë‚´ì—­ ê¸°ë¡
    // í•„ìš”ì‹œ Money Wave íŠ¸ë¦¬ê±°
  }
}

export class PmcAccountEventHandler {
  async handle(event: PmcEarnedEvent | PmcSpentEvent): Promise<void> {
    // PMC ê³„ì • ì”ì•¡ ì—…ë°ì´íŠ¸
    // ê±°ë˜ ë‚´ì—­ ê¸°ë¡
    // Money Wave ì¬ë¶„ë°° ì²˜ë¦¬
  }
}
```

### ë„ë©”ì¸ ê°„ ì´ë²¤íŠ¸ êµ¬ë…
```typescript
// ê° ë„ë©”ì¸ì€ ê´€ë ¨ ê²½ì œ ì´ë²¤íŠ¸ë¥¼ êµ¬ë…
export class PredictionEventHandler {
  async handleMoneyWaveDistribution(event: MoneyWaveDistributedEvent): Promise<void> {
    // ì˜ˆì¸¡ ê²Œì„ ì°¸ê°€ìë“¤ì—ê²Œ PMC ë¶„ë°°
  }
}
```

## TypeScript í‘œì¤€ (ì—…ë°ì´íŠ¸)

- ì—„ê²©í•œ TypeScript ì„¤ì • ì‚¬ìš©
- ê°ì²´ í˜•íƒœì—ëŠ” `type`ë³´ë‹¤ `interface` ì„ í˜¸
- ë„ë©”ì¸ ì‹ë³„ìì— ë¸Œëœë“œ íƒ€ì… ì‚¬ìš©:
```typescript
type UserId = string & { readonly brand: unique symbol };
type PredictionId = string & { readonly brand: unique symbol };
type PmpAmount = number & { readonly brand: unique symbol };
type PmcAmount = number & { readonly brand: unique symbol };
```
- ì—ëŸ¬ ì²˜ë¦¬ë¥¼ ìœ„í•œ Result/Either íŒ¨í„´ êµ¬í˜„:
```typescript
type Result<T, E = Error> = 
  | { success: true; data: T }
  | { success: false; error: E };
```

## ëª…ëª… ê·œì¹™

### íŒŒì¼ ë° ë””ë ‰í† ë¦¬
- íŒŒì¼: `kebab-case` â†’ `prediction-game.entity.ts`
- í´ë˜ìŠ¤: `PascalCase` â†’ `PredictionGame`
- í•¨ìˆ˜/ë³€ìˆ˜: `camelCase` â†’ `createPredictionGame`
- ìƒìˆ˜: `SCREAMING_SNAKE_CASE` â†’ `MAX_PREDICTIONS_PER_GAME`

### ë„ë©”ì¸ ê°ì²´
- ì—”í‹°í‹°: `PredictionGame`, `User`, `Market`
- ê°’ ê°ì²´: `PredictionId`, `UserId`, `Email`
- ë¦¬í¬ì§€í† ë¦¬: `IPredictionGameRepository` (ì¸í„°í˜ì´ìŠ¤), `PredictionGameRepository` (êµ¬í˜„)
- ì„œë¹„ìŠ¤: `PredictionScoringService`, `MarketDataService`
- ê²½ì œ ì„œë¹„ìŠ¤: `PredictionEconomicService`, `InvestmentEconomicService`

### React ì»´í¬ë„ŒíŠ¸
- ì»´í¬ë„ŒíŠ¸: `PredictionCard`, `GameList`, `UserDashboard`
- í›…: `usePredictionGame`, `useMarketData`, `useEconomicBalance`
- í˜ì´ì§€: `page.tsx`, `layout.tsx`

## ì½”ë“œ ìƒì„± ê·œì¹™

### ê²½ì œ í†µí•©ì´ í¬í•¨ëœ ë„ë©”ì¸ ì—”í‹°í‹° ìƒì„± ì‹œ
```typescript
export class PredictionGame {
  private constructor(
    private readonly id: PredictionGameId,
    private readonly title: string,
    private status: GameStatus,
    private predictions: Prediction[]
  ) {}

  public static create(title: string): Result<PredictionGame, ValidationError> {
    // ë„ë©”ì¸ ê²€ì¦ ë¡œì§
  }

  public addPrediction(
    prediction: Prediction,
    economicService: PredictionEconomicService
  ): Result<void, BusinessRuleError> {
    // ê²½ì œì  ê²€ì¦ì´ í¬í•¨ëœ ë¹„ì¦ˆë‹ˆìŠ¤ ë¡œì§
  }
}
```

### ë¦¬í¬ì§€í† ë¦¬ ì¸í„°í˜ì´ìŠ¤ ìƒì„± ì‹œ
```typescript
// ë„ë©”ì¸ ê³„ì¸µ - êµ¬í˜„ ì„¸ë¶€ì‚¬í•­ ì—†ìŒ
export interface IPredictionGameRepository {
  save(game: PredictionGame): Promise<Result<void, RepositoryError>>;
  findById(
    id: PredictionGameId
  ): Promise<Result<PredictionGame | null, RepositoryError>>;
  findByStatus(
    status: GameStatus
  ): Promise<Result<PredictionGame[], RepositoryError>>;
}
```

### ê²½ì œ í†µí•©ì´ í¬í•¨ëœ ìœ ìŠ¤ì¼€ì´ìŠ¤ ìƒì„± ì‹œ
```typescript
// ì• í”Œë¦¬ì¼€ì´ì…˜ ê³„ì¸µ - ë„ë©”ì¸ ê°ì²´ì™€ ê²½ì œ ì»¤ë„ ì¡°ìœ¨
export class CreatePredictionGameUseCase {
  constructor(
    private readonly repository: IPredictionGameRepository,
    private readonly eventPublisher: IDomainEventPublisher,
    private readonly economicService: PredictionEconomicService
  ) {}

  async execute(
    request: CreateGameRequest
  ): Promise<Result<CreateGameResponse, UseCaseError>> {
    // 1. ì…ë ¥ ê²€ì¦
    // 2. ê²½ì œì  ì „ì œ ì¡°ê±´ í™•ì¸
    // 3. ë„ë©”ì¸ ê°ì²´ ìƒì„±
    // 4. ë¦¬í¬ì§€í† ë¦¬ë¥¼ í†µí•´ ì €ì¥
    // 5. ë„ë©”ì¸ ì´ë²¤íŠ¸ ë°œí–‰ (ê²½ì œ ì´ë²¤íŠ¸ í¬í•¨)
    // 6. ì‘ë‹µ ë°˜í™˜
  }
}
```

### ê²½ì œ ë°ì´í„°ê°€ í¬í•¨ëœ React ì»´í¬ë„ŒíŠ¸ ìƒì„± ì‹œ
- ê¸°ë³¸ì ìœ¼ë¡œ Server Component ì‚¬ìš©
- í•„ìš”í•œ ê²½ìš°ì—ë§Œ Client Component ì‚¬ìš© (ìƒí˜¸ì‘ìš©, ë¸Œë¼ìš°ì € API)
- Suspenseë¡œ ì ì ˆí•œ ë¡œë”© ìƒíƒœ êµ¬í˜„
- ì—ëŸ¬ ê²½ê³„ í¬í•¨
- **ê´€ë ¨ ì»´í¬ë„ŒíŠ¸ì—ì„œ í•­ìƒ ê²½ì œ ì”ì•¡(PMP/PMC) í‘œì‹œ**
```typescript
// ê²½ì œ ë°ì´í„°ê°€ í¬í•¨ëœ Server Component
export default async function GameList({ userId }: { userId: UserId }) {
  const [games, pmpBalance, pmcBalance] = await Promise.all([
    getGames(),
    getPmpBalance(userId),
    getPmcBalance(userId)
  ]);
  
  return (
    <div>
      <EconomicBalance pmp={pmpBalance} pmc={pmcBalance} />
      {/* ê²Œì„ ëª©ë¡ JSX */}
    </div>
  );
}

// Client Component (í•„ìš”í•œ ê²½ìš°)
("use client");
export function PredictionForm({ userId }: { userId: UserId }) {
  const { pmpBalance } = useEconomicBalance(userId);
  // ê²½ì œì  ê²€ì¦ì´ í¬í•¨ëœ ìƒí˜¸ì‘ìš© í¼ ë¡œì§
}
```

## Supabase í†µí•© ê·œì¹™

### ê²½ì œ í†µí•©ì„ ìœ„í•œ ë°ì´í„°ë² ì´ìŠ¤ ìŠ¤í‚¤ë§ˆ
- ë°ì´í„°ë² ì´ìŠ¤ ì»¬ëŸ¼ì—ëŠ” snake_case ì‚¬ìš©
- ë„ë©”ì¸ ê°ì²´ì—ì„œëŠ” camelCaseë¡œ ë§¤í•‘
- Row Level Security (RLS) êµ¬í˜„
- **ê²½ì œ ê±°ë˜ì™€ ë„ë©”ì¸ ì—”í‹°í‹°ë¥¼ ìœ„í•œ ë³„ë„ í…Œì´ë¸”**
```sql
-- ê²½ì œ í…Œì´ë¸” (ê³µìœ )
CREATE TABLE pmp_accounts (
  user_id UUID PRIMARY KEY,
  balance DECIMAL(15,2) NOT NULL DEFAULT 0,
  updated_at TIMESTAMPTZ DEFAULT NOW()
);

CREATE TABLE pmc_accounts (
  user_id UUID PRIMARY KEY,
  balance DECIMAL(15,2) NOT NULL DEFAULT 0,
  updated_at TIMESTAMPTZ DEFAULT NOW()
);

CREATE TABLE economic_transactions (
  id UUID PRIMARY KEY,
  user_id UUID NOT NULL,
  transaction_type TEXT NOT NULL, -- 'pmp_earned', 'pmp_spent', 'pmc_earned', 'pmc_spent'
  amount DECIMAL(15,2) NOT NULL,
  source_domain TEXT NOT NULL, -- 'prediction', 'investment', 'forum', 'donation'
  source_id UUID NOT NULL,
  created_at TIMESTAMPTZ DEFAULT NOW()
);

-- ë„ë©”ì¸ë³„ í…Œì´ë¸”
CREATE TABLE prediction_games (
  id UUID PRIMARY KEY,
  created_at TIMESTAMPTZ DEFAULT NOW(),
  game_status TEXT NOT NULL,
  required_pmp DECIMAL(15,2) DEFAULT 0
);
```

### ê²½ì œ í†µí•©ì´ í¬í•¨ëœ ë¦¬í¬ì§€í† ë¦¬ êµ¬í˜„
```typescript
export class SupabasePredictionGameRepository
  implements IPredictionGameRepository
{
  constructor(private readonly supabase: SupabaseClient) {}

  async save(game: PredictionGame): Promise<Result<void, RepositoryError>> {
    try {
      // ë„ë©”ì¸ ê°ì²´ë¥¼ ë°ì´í„°ë² ì´ìŠ¤ í˜•ì‹ìœ¼ë¡œ ë³€í™˜
      // Supabase í´ë¼ì´ì–¸íŠ¸ ì‚¬ìš©
      // ì ì ˆí•œ ì—ëŸ¬ ì²˜ë¦¬
    } catch (error) {
      return { success: false, error: new RepositoryError("ì €ì¥ ì‹¤íŒ¨") };
    }
  }
}
```

## ì—ëŸ¬ ì²˜ë¦¬ íŒ¨í„´

### ë„ë©”ì¸ ì—ëŸ¬
```typescript
export class ValidationError extends Error {
  constructor(message: string, public field: string) {
    super(message);
    this.name = "ValidationError";
  }
}

export class BusinessRuleError extends Error {
  constructor(message: string, public rule: string) {
    super(message);
    this.name = "BusinessRuleError";
  }
}

export class EconomicError extends Error {
  constructor(
    message: string,
    public economicType: 'insufficient-pmp' | 'insufficient-pmc' | 'invalid-transaction'
  ) {
    super(message);
    this.name = "EconomicError";
  }
}
```

### ì• í”Œë¦¬ì¼€ì´ì…˜ ì—ëŸ¬
```typescript
export class UseCaseError extends Error {
  constructor(message: string, public cause?: Error) {
    super(message);
    this.name = "UseCaseError";
  }
}
```

## Next.js 15 App Router ê·œì¹™

- App Router íŒŒì¼ ê·œì¹™ ì‚¬ìš©
- loading.tsx, error.tsx, not-found.tsx êµ¬í˜„
- ë³€ê²½ ì‘ì—…ì— Server Actions í™œìš©
- ì ì ˆí•œ ìºì‹± ì „ëµ ì‚¬ìš©
- **í˜ì´ì§€ propsì— í•­ìƒ ê²½ì œ ë°ì´í„° í¬í•¨**
```typescript
// app/predictions/loading.tsx
export default function Loading() {
  return <PredictionsSkeleton />;
}

// app/predictions/error.tsx
("use client");
export default function Error({
  error,
  reset,
}: {
  error: Error & { digest?: string };
  reset: () => void;
}) {
  return <ErrorBoundary error={error} reset={reset} />;
}
```

## PowerShell ëª…ë ¹ì–´ (í•­ìƒ ì‚¬ìš©)

```powershell
# ê°œë°œ
npm run dev
npm test
npm run build

# ì—¬ëŸ¬ ëª…ë ¹ì–´ (ì„¸ë¯¸ì½œë¡  ì‚¬ìš©)
cd src; npm install; npm run dev

# íŒŒì¼ ì‘ì—…
New-Item -ItemType Directory -Path "src\bounded-contexts\prediction"
```

## ì„±ëŠ¥ ë° ë³´ì•ˆ

### ì„±ëŠ¥
- ë¹„ìš©ì´ ë§ì´ ë“œëŠ” ì»´í¬ë„ŒíŠ¸ì— React.memo ì‚¬ìš©
- ì ì ˆí•œ ë°ì´í„°ë² ì´ìŠ¤ ì¸ë±ì‹± êµ¬í˜„
- ì½”ë“œ ë¶„í• ì„ ìœ„í•œ ë™ì  ì„í¬íŠ¸ ì‚¬ìš©
- ìºì‹± ì „ëµ êµ¬í˜„
- **ì ì ˆí•œ ë¬´íš¨í™”ë¥¼ í†µí•œ ê²½ì œ ì”ì•¡ ë°ì´í„° ìºì‹±**

### ë³´ì•ˆ
- API ê²½ê³„ì—ì„œ ëª¨ë“  ì…ë ¥ ê²€ì¦
- ë°ì´í„° ì ‘ê·¼ ì œì–´ë¥¼ ìœ„í•œ Supabase RLS ì‚¬ìš©
- í™˜ê²½ ë³€ìˆ˜ì— ë¯¼ê°í•œ ë°ì´í„° ì €ì¥
- OWASP ë³´ì•ˆ ê°€ì´ë“œë¼ì¸ ì¤€ìˆ˜
- **ê²½ì œ ê±°ë˜ ê²€ì¦ ë° ê°ì‚¬ ì¶”ì  êµ¬í˜„**

## PosMul ë„ë©”ì¸ë³„ ê·œì¹™

### ê²½ì œ ë„ë©”ì¸ (ê³µìœ  ì»¤ë„)
- PMP (ë¬´ìœ„í—˜ ìì‚°)ì™€ PMC (ìœ„í—˜ ìì‚°)ëŠ” í•µì‹¬ ê°’ ê°ì²´
- MoneyWave ì‹œìŠ¤í…œì´ PMC ë¶„ë°° ì²˜ë¦¬
- ëŒ€ë¦¬ì¸ ì´ë¡  êµ¬í˜„ì´ í•µì‹¬ ì°¨ë³„í™” ìš”ì†Œ
- **ëª¨ë“  ê²½ì œ ì‘ì—…ì€ ë„ë©”ì¸ ì´ë²¤íŠ¸ë¥¼ í†µí•´ ì²˜ë¦¬**
- **ê²½ì œ ë„ë©”ì¸ ì™¸ë¶€ì—ì„œ ì§ì ‘ ì”ì•¡ ìˆ˜ì • ì ˆëŒ€ ê¸ˆì§€**

### ì˜ˆì¸¡ ë„ë©”ì¸
- PredictionGameì´ ì£¼ìš” ì• ê·¸ë¦¬ê±°íŠ¸
- Predictionì€ ì• ê·¸ë¦¬ê±°íŠ¸ ë‚´ ì—”í‹°í‹°
- ê²Œì„ ìƒëª…ì£¼ê¸°: ìƒì„±ë¨ â†’ í™œì„± â†’ ì¢…ë£Œë¨ â†’ ì •ì‚°ë¨
- **ì°¸ì—¬ í—ˆìš© ì „ í•­ìƒ PMP ì”ì•¡ í™•ì¸**
- **ì„±ê³µì ì¸ ì˜ˆì¸¡ ì‹œ PmcEarnedEvent ë°œí–‰**

### íˆ¬ì ë„ë©”ì¸
- Local League, Major League, Cloud Fundingì€ ë³„ë„ ì• ê·¸ë¦¬ê±°íŠ¸
- ê°ê° ë‹¤ë¥¸ í¬ì¸íŠ¸ íšë“ ë©”ì»¤ë‹ˆì¦˜ ë³´ìœ 
- **Major LeagueëŠ” PMP íšë“, Local LeagueëŠ” PMC íšë“**
- **Cloud Fundingì€ PMC íšë“**

### í¬ëŸ¼ ë„ë©”ì¸
- **í† ë¡ ê³¼ ë¸Œë ˆì¸ìŠ¤í† ë°ì€ PMP íšë“**
- **ì–‘ì§ˆì˜ ê¸°ì—¬ëŠ” ê²½ì œì  ì¸ì„¼í‹°ë¸Œë¡œ ë³´ìƒ**

### ê¸°ë¶€ ë„ë©”ì¸
- **ê¸°ë¶€ì— PMC ì‚¬ìš©**
- **Money Wave ì¬ë¶„ë°° ë¡œì§ êµ¬í˜„**

## ì¤‘ìš” ì£¼ì˜ì‚¬í•­

1. **í´ë¦° ì•„í‚¤í…ì²˜ ì˜ì¡´ì„± ê·œì¹™ ì ˆëŒ€ ìœ„ë°˜ ê¸ˆì§€**
2. **ë„ë©”ì¸ ê³„ì¸µì€ ìˆœìˆ˜í•´ì•¼ í•¨** - ì™¸ë¶€ ì˜ì¡´ì„± ì—†ìŒ
3. **PowerShell ëª…ë ¹ì–´ ì‚¬ìš©** - bash/zsh ì ˆëŒ€ ê¸ˆì§€
4. **í•­ìƒ Result íŒ¨í„´ êµ¬í˜„** ì—ëŸ¬ ì²˜ë¦¬ìš©
5. **ë¦¬í¬ì§€í† ë¦¬ ì¸í„°í˜ì´ìŠ¤ëŠ” ë„ë©”ì¸ì—** - êµ¬í˜„ì€ ì¸í”„ë¼ìŠ¤íŠ¸ëŸ­ì²˜ì—
6. **ê¸°ë³¸ì ìœ¼ë¡œ Server Components** - í•„ìš”ì‹œì—ë§Œ Client
7. **ì ì ˆí•œ ì—ëŸ¬ ê²½ê³„** ë° ë¡œë”© ìƒíƒœ
8. **ëª¨ë“  ë°ì´í„° ì ‘ê·¼ì— Supabase RLS**
9. **ğŸ”¥ ê²½ì œ í†µí•©ì´ ì¤‘ìš”** - ëª¨ë“  ë„ë©”ì¸ì´ PMP/PMC ì‹œìŠ¤í…œê³¼ í†µí•©ë˜ì–´ì•¼ í•¨
10. **ë„ë©”ì¸ ê°„ ê²½ì œ ê±°ë˜ì— ë„ë©”ì¸ ì´ë²¤íŠ¸ ì‚¬ìš©**
11. **ì½ê¸° ì ‘ê·¼ì€ ê³µìœ  ê²½ì œ ì»¤ë„, ì“°ê¸° ì ‘ê·¼ì€ ì´ë²¤íŠ¸ë¥¼ í†µí•´**
12. **ê´€ë ¨ UI ì»´í¬ë„ŒíŠ¸ì—ì„œ í•­ìƒ ê²½ì œ ì”ì•¡ í‘œì‹œ**

ì˜ì‹¬ìŠ¤ëŸ¬ìš¸ ë•ŒëŠ” ìŠ¤ìŠ¤ë¡œì—ê²Œ ë¬¼ì–´ë³´ì„¸ìš”: "ì´ê²ƒì´ DDD ì›ì¹™, í´ë¦° ì•„í‚¤í…ì²˜ ê·œì¹™ì„ ë”°ë¥´ê³ , ê²½ì œ ì‹œìŠ¤í…œê³¼ ì ì ˆíˆ í†µí•©ë˜ëŠ”ê°€?" 